src/tools/pub/publishing.md

```diff
@@ -140,6 +140,38 @@ the first and only person authorized to upload additional versions of the packag
 To allow or disallow other people to upload versions,
 use the [`pub uploader`](cmd/pub-uploader) command.
 
+## Publishing pre-releases
+
+As you work on a package, consider publishing it as a pre-release.
+Pre-releases can be useful when any of the following are true:
+
+* You're actively developing the next major version of the package.
+* You want beta testers for the next release candidate of the package.
+* The package depends on an unstable version of the Dart or Flutter SDK.
+
+As described in [semantic versioning][semver], to make a pre-release of a version,
+you append a suffix to the version. For example, to make a pre-release of
+version `2.0.0` you might use the version `2.0.0-dev.1`. Later, when you
+release version `2.0.0`, it will take precedence over all `2.0.0-XXX` pre-releases.
+
+Because `pub` prefers stable releases when available, users of a pre-release package
+might need to change their dependency constraints.
+For example, if a user wants to test pre-releases of version 2.1, then
+instead of `^2.0.0` or `^2.1.0` they might specify `^2.1.0-dev.1`.
+
+<aside class="alert alert-info" markdown="1">
+  **Note:**
+  Sometimes `pub` might choose a pre-release instead of a stable release;
+  this can occur when dependencies are nested.
```

Comment:
I wish we could say something more precise here :)

I did a bit of digging and filed some tests in: https://github.com/dart-lang/pub/pull/2175

I think the sane and simple thing we can say something along the lines of:
> `pub` may choose a pre-release instead of a stable release, if there exists a stable package in the dependency graph depends on a pre-release.

(I'm sure it can be phrased better)

@jakemac53, isn't that a reasonably fair characterization of the cases where pre-releases may be chosen, despite stable releases being available?

Reply:
Yes, that matches my understanding pretty well.

Reply:
Well, actually it is more complicated than that but I don't know that it matters? There doesn't need to be any explicit dependencies on a pre-release. It can happen if a pre-release is the only version that allows the newest version of some other stable dependency.